Upstream apply_optim_in_backward from TorchRec

rohan-varma opened the pull request:
Summary:
Upstreaming this as part of sharing common APIs. This is just a plain
move, any changes needed to support DDP / FSDP will come in follow up diffs.

Test Plan: CI

Differential Revision: D40564646

colin2328 commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,80 @@
+        if hasattr(param, "_acc_grad") and hasattr(param, "_overlapped_optimizer"):
+            raise ValueError(
+                # pyre-ignore
+                f"{param} already has {param._overlapped_optimizer} applied as an overlapped optimizer. Cannot apply again"
```
can we call this `_in_backward_optimizer` instead of `_overlapped_optimizer` to be consistent?

colin2328 commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+        optimizer = optimizer_class([param], **optimizer_kwargs)
+
+        if (
+            hasattr(param, "_acc_grad") and hasattr(param, "_in_backward_optimizer")  # type: ignore[attr-defined]
```
should we support more than one optimizer per param in the future? (perhaps through CombinedOptimizer or different implementation)
perhaps add a TODO here for now

rohan-varma replied:
Yep, we should eventually be able to support more than one opt per param. @zhaojuanmao mentioned an internal use case that might need this as well. Will add TODO / issue

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env python3
```
Curious: Do we need to include this?

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+    a no-op.
+
+    Args:
+        optimizer_class: Type[torch.optim.Optimizer]: Optimizer to apply to parameter
```
Shall we check that these render as expected?

rohan-varma replied:
I am planning to make this private for now, since this is a prototype API.

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+    """
+
+    def _apply_optimizer_in_backward_to_param(param: torch.nn.Parameter) -> None:
+        # acc_grad creates a new node in the auto_grad graph that comes after
```
I think we should reword this comment.

`view_as()` is what creates a node that gives us access to the parameter's `AccumulateGrad` object.

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+    def _apply_optimizer_in_backward_to_param(param: torch.nn.Parameter) -> None:
+        # acc_grad creates a new node in the auto_grad graph that comes after
+        # the parameter node. In this node's backwards we call the overlapped
+        # optimizer.step(). We cannot do the backwards hook on param because
```
I think what this is trying to explain is that we cannot register the backward hook on the parameter itself. It must be on the parameter's `AccumulateGrad` object.

This is indeed my understanding, but I think this can be stated more clearly. The point is that registering a hook on the parameter means that hook runs before its gradient computation, while registering a hook on the `AccumulateGrad` object means that hook runs after the gradient computation.

rohan-varma replied:
I clarified this in next update.

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+        ):
+            raise ValueError(
+                # pyre-ignore
+                f"{param} already has {param._in_backward_optimizer} "  # type: ignore[attr-defined]
```
Curious: Will PyTorch truncate the `param` when converting to string? Otherwise, I wonder if this will print a massive tensor.

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+            )
+
+        # The grad accumulator is a weak ref, so we need to keep it
+        # alive until the Tensor is alive.
```
"until the Tensor is alive"?

Which tensor? This also makes it sound like the tensor is not yet alive.

rohan-varma replied:
Fixed

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+        param._acc_grad = acc_grad  # type: ignore[attr-defined]
+        param._in_backward_optimizer = optimizer  # type: ignore[attr-defined]
+
+        # pyre-ignore
```
I think we can remove these `pyre-ignore`s for Github.

awgu commented on test/distributed/optim/test_apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,57 @@
+        for i in range(6):
+            inp = torch.randn(4, 10)
+            model(inp).sum().backward()
+            model_with_opt_in_bwd(inp.clone()).sum().backward()
```
Do we need to `clone()`? If so, why?

awgu commented on test/distributed/optim/test_apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,57 @@
+            model(inp).sum().backward()
+            model_with_opt_in_bwd(inp.clone()).sum().backward()
+            optim.step()
+            with self.subTest(i):
```
I am guessing this is for reporting that the test fails in iter `i`?

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+    Example::
+        params_generator = model.parameters()
+        param_1 = next(params_generator)
+        param_2 = list(params_generator)
```
nit: I am not a fan of this variable name `param_2` when it is actually a `list`.

rohan-varma replied:
Changed this.

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+        apply_optimizer_in_backward(torch.optim.SGD, [param_1], {"lr": .02})
+        apply_optimizer_in_backward(torch.optim.Adam, param_2, {"lr": .04})
+
+        print(param_1._optimizer, param_1._optimizer_kwargs)
```
Are we meaning to advertise this private `_optimizer` attribute since we are including it in this example code snippet?

rohan-varma replied:
Right, we should probably not advertise this as such.

zhaojuanmao commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+            param._in_backward_optimizer.step()  # type: ignore[attr-defined]
+            param.grad = None
+
+        param._acc_grad.register_hook(optimizer_hook)  # type: ignore[attr-defined]
```
this optimizer_hook will mostly be fired before other auto grad post hooks installed by developers (like DDP & FSDP) or users if users call apply_optimizer_in_backward() before model wrapping?

zhaojuanmao commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,85 @@
+        if (
+            hasattr(param, "_acc_grad") and hasattr(param, "_in_backward_optimizer")  # type: ignore[attr-defined]
+        ):
+            raise ValueError(
```
just to confirm that shared parameters will have different "_in_backward_optimizer" and "_acc_grad"?

rohan-varma replied:
I think they might override / throw an error. Will add a test.

rohan-varma replied:
Yes, this currently throws an error. Typically, if the user has the following:

m = nn.Linear(...)
m.weight = m.bias # share params
weight_opt = torch.optim.SGD([m.weight])
bias_opt = torch.optim.SGD([m.bias])

then calling .step() will result in 2 weight updates both to the same underlying tensor. So this is the behavior we should follow.

rohan-varma replied:
Fixed to just include multiple optimizers.

awgu commented:
Let me review one more time later today.

awgu commented on torch/distributed/optim/__init__.py:
```diff
@@ -17,6 +17,7 @@
 from .functional_rprop import _FunctionalRprop
 from .functional_adamax import _FunctionalAdamax
 from .utils import as_functional_optim
+from .apply_optimizer_in_backward import _apply_optimizer_in_backward
```
nit: Does this need to be sorted?

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+
+import torch
+
+__all__ = []  # type: ignore[var-annotated]
```
nit: I wonder if we just annotated this as `__all__: List[str] = []`, we can remove the `# type: ignore[var-annotated]`.

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+    optimizer_kwargs: Dict[str, Any],
+) -> None:
+    """
+    Upon backwards(), parameters will fire the corresponding optimizer.
```
nit: The backtick is a super nit, but the important thing is that the method is `backward()`.
```suggestion
    Upon ``backward()``, parameters will fire the corresponding optimizer.
```

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+    """
+    Upon backwards(), parameters will fire the corresponding optimizer.
+
+    Note - gradients for these parameters will be set to None after backwards().
```
nit: same here
```suggestion
    Note - gradients for these parameters will be set to ``None`` after ``backward()``.
```

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+    a no-op.
+
+    Args:
+        optimizer_class: Type[torch.optim.Optimizer]: Optimizer to apply to parameter
```
nit:
```suggestion
        optimizer_class (Type[torch.optim.Optimizer]): Optimizer to apply to parameter
```

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+
+    Args:
+        optimizer_class: Type[torch.optim.Optimizer]: Optimizer to apply to parameter
+        params: Iterator[nn.Parameter]: parameters to apply optimizer state to
```
nit:
```suggestion
        params (Iterator[nn.Parameter]): parameters to apply optimizer state to
```

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+    Args:
+        optimizer_class: Type[torch.optim.Optimizer]: Optimizer to apply to parameter
+        params: Iterator[nn.Parameter]: parameters to apply optimizer state to
+        optimizer_kwargs: Dict[str, Any]: kwargs to pass to optimizer constructor
```
nit:
```suggestion
        optimizer_kwargs (Dict[str, Any]): kwargs to pass to optimizer constructor
```

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+    Example::
+        params_generator = model.parameters()
+        param_1 = next(params_generator)
+        remainder_params = list(params_generator)
```
nit: maybe?
```suggestion
        remaining_params = list(params_generator)
```

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+
+    """
+
+    def _apply_optimizer_in_backward_to_param(param: torch.nn.Parameter) -> None:
```
Maybe we should just add `@no_type_check`?

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+        # hook on this object to fire the optimizer when the gradient for
+        # this parameter is ready (has been accumulated into .grad field)
+
+        # Don't create a new acc_grad if we already have one (i.e. shared params).
```
"i.e. shared params" -> Is this also relevant for multiple optimizers per parameter like if we call this `_apply_optimizers_in_backward_to_param()` multiple times?

Also, for FSDP, we also save the hook itself. Would this ever be necessary here?

rohan-varma replied:
Yes. will add that as a docstring.

We will probably need to eventually save the hook, for example so that DDP can remove it.

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+
+    """
+
+    def _apply_optimizer_in_backward_to_param(param: torch.nn.Parameter) -> None:
```
Do we want to call it `_apply_optimizer_in_backward_to_param()` or simply just `_apply_optimizer_in_backward()`?

I might have missed some naming discussion. If we go with the former, then we should update the docstring example.

rohan-varma replied:
The overall function is called _apply_optimizer_in_backward, this is an inner helper method that applies optimizer to a single param.

awgu replied:
Ah, makes sense!

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+
+            param.grad = None
+
+        param._acc_grad.register_hook(optimizer_hook)  # type: ignore[attr-defined]
```
I am concerned that this hook may be registered multiple times if we call this method multiple times. This might be why we should save the hook object, remove it, and re-register it, or we may not need to register it again at all.

I have not though it completely through, but we should just clarify what is the invariant here. For example, `hasattr(param, "_acc_grad")` iff hook registered? Then, we should make sure not to re-register the hook? It seems fine to not re-register because the hook should already have the necessary data structure references in its closure.

awgu reviewed (commented):
I left a lot of nits, but I had some actual questions about the logic.

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -0,0 +1,76 @@
+        optimizer = optimizer_class([param], **optimizer_kwargs)
+
+        # Keep the grad accumulator around for the lifetime of the Tensor,
+        # store it on the module to avoid uncollectable ref-cycle
```
"store it on the module to avoid uncollectable ref-cycle"

This seems to be stored on the `param`?

awgu commented:
@pytorchbot merge

rohan-varma committed:
Upstream apply_optim_in_backward from TorchRec (#87397)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/87397

Upstreaming this as part of sharing common APIs. This is just a plain
move, any changes needed to support DDP / FSDP will come in follow up diffs.

Test Plan: CI

Reviewed By: zhaojuanmao

Differential Revision: D40564646

fbshipit-source-id: 619c434e02196812f8d4db1e40d07290e08b18f9
